SPARC: Do not pass g0 to zeroflags

mumbel opened the pull request:
RD cannot be used with the zeroflags() macro as it could be hardcoded zero g0 register

mumbel commented on Ghidra/Processors/Sparc/data/languages/SparcV9.sinc:
```diff
@@ -877,48 +885,53 @@ callreloff: reloc  is disp30 [reloc=inst_start+4*disp30;] { export *:$(SIZE) rel
-	denom:$(SIZE) = regorimm & 0xffffffff;
-	RD = numerator / denom;
+	numerator:8 = (zext(Y) << 32) + zext(RS1);
+	denom:8 = zext(regorimm:4)
```
```suggestion
	denom:8 = zext(regorimm:4);
```

mumbel commented on Ghidra/Processors/Sparc/data/languages/SparcV9.sinc:
```diff
@@ -877,48 +885,53 @@ callreloff: reloc  is disp30 [reloc=inst_start+4*disp30;] { export *:$(SIZE) rel
-	zeroflags(RD);
-	i_vf = RD > 0xffffffff;
+	numerator:8 = (zext(Y) << 32) + zext(RS1:4);
+	denom:8 = zext(regorimm:4)
```
```suggestion
	denom:8 = zext(regorimm:4);
```

mumbel commented:
@ghidracadabra (also @emteere since I think I've seen you do this, riscv maybe) opinions? Should `sw_trap` be a read of  trap address and a `call`

I saw something in 11.0.2 release notes seemingly related, but I have not tried that release yet

emteere reviewed (commented):
For the patterns, since the "save sp" Is such a strong pattern.
Instead of enumerating all the filler patterns, I think you can just make a single pre-pattern with
```
<data>0x00000000 0x00000000 </data> <!-- RETURN : delayslot filler/nop -->
```

And then each of the more unique two word end patterns
```
<data>01...... 0x...... 10.....1 11101... 0x.... </data> <!-- CALL label; RESTORE : filler/nop -->
<data>0x00 101..... 0x.... </data> <!-- BA,A -label : filler/nop -->
<data>0x81cfe008 0x........ </data> <!-- RETURN : delayslot filler/nop -->
....
....
```

mumbel commented:
@emteere   I had just used the template that was there before and it did look kind of weird.  I've rebased/squashed, and changed that file... is that what you meant?

emteere commented on Ghidra/Processors/Sparc/data/languages/SparcV9.sinc:
```diff
@@ -825,51 +835,49 @@ callreloff: reloc  is disp30 [reloc=inst_start+4*disp30;] { export *:$(SIZE) rel
-:smul    RS1,regorimm,RD             is op=2 & RD & op3=0x0b & RS1 & regorimm {RD = sext(RS1:4) * sext(regorimm:4); Y=RD>>32;}
-:umulcc  RS1,regorimm,RD             is op=2 & RD & op3=0x1a & RS1 & regorimm {RD = zext(RS1:4) * zext(regorimm:4); Y=RD>>32; zeroflags(RD); logicflags();}
-:smulcc  RS1,regorimm,RD             is op=2 & RD & op3=0x1b & RS1 & regorimm {RD = sext(RS1:4) * sext(regorimm:4); Y=RD>>32; zeroflags(RD); logicflags();}
+:umul    RS1,regorimm,RD             is op=2 & RD & op3=0x0a & RS1 & regorimm {local res:8 = zext(RS1:4) * zext(regorimm:4); Y=res >> 32; RD = res:4;}
```
Prefer separate lines for pcode semantics.

emteere commented:
For the Tcc instruction sw_trap(), I do think it should be changed to branch to an address, or maybe call to a computed address from the TBA table.
It is also missing an if condition on the trap condition code.
Is that what you were referring to in an earlier comment?

This probably deserves a separate ticket for this change.  Especially if this PR is going to get into 11.2.

mumbel commented on Ghidra/Processors/Sparc/data/patterns/SPARC_patterns.xml:
```diff
@@ -2,30 +2,21 @@
+      <data>0x81cfe008 0x........ 0000000. 0x000000 </data> <!-- RETURN :  delayslot filler/nop -->
+      <data>0x00 101..... 0x.... 0x........ </data> <!-- BA -label :  delayslot -->
+      <data>0x00 101..... 0x.... 0x........ 0000000. 0x000000 </data> <!-- BA -label :  delayslot filler/nop -->
+      <data>0x00 101..... 0x.... </data> <!-- BA,A -label :  filler/nop -->
```
This should mean no delayslot, I'll double check and update if needed, but thought that's what the ",a" allowed for

mumbel replied:
i messed this up,  0x00.... should have been 0x10 or 0x30.  thanks

mumbel commented on Ghidra/Processors/Sparc/data/languages/SparcV9.sinc:
```diff
@@ -990,6 +1003,12 @@ sethidisp: "%hi("^hi^")"  is udisp22 [hi=udisp22<<10;] { export *[const]:$(SIZE)
 :restore  RS1,regorimm,RD            is op=0x2 & RD & op3=0x3d & RS1 & regorimm { local tmp = RS1 + regorimm; restore(); didrestore=1; RD = tmp; }
 :restore                             is op=0x2 & rd=0 & op3=0x3d                {                             restore(); didrestore=1; }
 
+# FIXME  'jmpl' can have 'return' in the delayslot to return from a user trap handler
```
This is #6285   I didn't know how that would be best handled in code.  A fix isn't obvious, would it be best to remove this comment and just have the GitHub issue, reference the GitHub issue here, or just add more details/example here

mumbel commented:
Thanks for the review, just saw the email.  I'll try to address tonight

mumbel commented on Ghidra/Processors/Sparc/data/patterns/SPARC_patterns.xml:
```diff
@@ -2,30 +2,21 @@
+      <data>0x81c3e008 0x........ 0000000. 0x000000 </data> <!-- RETL :  delayslot filler/nop -->
+      <data>0x81cfe008 0x........ </data> <!-- RETURN :  delayslot -->
+      <data>0x81cfe008 0x........ 0000000. 0x000000 </data> <!-- RETURN :  delayslot filler/nop -->
+      <data>0x00 101..... 0x.... 0x........ </data> <!-- BA -label :  delayslot -->
```
ooo... yeah. that is a bad copy paste looks like and i broke that pattern. thanks

mumbel commented:
> For the Tcc instruction sw_trap(), I do think it should be changed to branch to an address, or maybe call to a computed address from the TBA table. It is also missing an if condition on the trap condition code. Is that what you were referring to in an earlier comment?
> 
> This probably deserves a separate ticket for this change. Especially if this PR is going to get into 11.2.

yes.  I agree it might be more complex than i was thinking originally and just do another PR for the address table lookup.

I do think the `if (!icc) goto inst_next;`  is reasonable in this PR though, I didn't see that, thanks.

emteere commented on Ghidra/Processors/Sparc/data/languages/SparcV9.sinc:
```diff
@@ -825,51 +835,73 @@ callreloff: reloc  is disp30 [reloc=inst_start+4*disp30;] { export *:$(SIZE) rel
 
 #----------------MULTIPLY 32 bit
 @if SIZE=="8"
```
I'm going to collapse the 32/64 bit variants.  Now that you have fixed the implementation, they perform the same operation.  Only difference is the size of the Y register

emteere commented on Ghidra/Processors/Sparc/data/languages/SparcV9.sinc:
```diff
@@ -1070,7 +1131,11 @@ TICC:  "%xcc"  is cc1_4=1 &cc0_4=0 { }
 trap: RS1+swtrap  is i=1 & RS1 & swtrap { local tmp = ((RS1 + swtrap) & 0x7F); export tmp; }
 
-:t^tcc TICC, trap                    is op=0x2 & op3=0x3a & tcc & TICC & trap { sw_trap(trap); }
+:t^tcc TICC, trap                    is op=0x2 & op3=0x3a & tcc & TICC & trap
```
Adding a slight modification to the trap so that it can be over-ridden not to return for a trap.
```
:t^tcc TICC, trap                    is op=0x2 & op3=0x3a & tcc & TICC & trap
{
	if (!tcc) goto inst_next;
	local target:$(SIZE) = sw_trap(trap);
	call [target];
}
```

emteere commented:
Thanks for the PR!  Should be merged in soon, with a few minor changes.

mumbel commented:
Appreciate the help getting it in shape/merged
